REGRESSION(265043@main): Fullscreen mode only shows video in some part of the screen on Twitter

nt1m opened the pull request:
#### 5198e14b51f73c112c7f28c80558a9e491190810
<pre>
REGRESSION(265043@main): Fullscreen mode only shows video in some part of the screen on Twitter
<a href="https://bugs.webkit.org/show_bug.cgi?id=258907">https://bugs.webkit.org/show_bug.cgi?id=258907</a>
rdar://111095697

Reviewed by NOBODY (OOPS!).

265043@main started to use container queries in the video shadow root. As a side effect, this caused every call to
Document::resolveStyle() to start laying out the whole document (through frameView.layoutContext().layout()).

Adding to top layer &amp; removing from top layer requires layout when the containing block changes.

We haven&apos;t really hit any problem with not marking for layout until now, because usually we apply the pseudo element styles
which include `position: fixed;` and other things, at the same time as adding to top layer, which marks them for layout anyway.

In the fullscreen case, we have a slightly different series:
```
ancestor-&gt;setFullscreenFlag(true); &lt;- applies pseudo class styles (hence dirtying layout)

ancestor-&gt;document().resolveStyle(Document::ResolveStyleType::Rebuild); &lt;- runs layout (clearing dirty layout)

ancestor-&gt;addToTopLayer(); &lt;- changes the containing block without marking renderer for layout

// No layout!!
```

Add a special case in addToTopLayer/removeFromTopLayer, for cases where we don&apos;t process pseudo class style changes and containing block changes
under the same layout.

Also make it clear in the media-controls CSS that container queries are needed on all platforms (not just visionOS), since
WebVTT relies on it.

* LayoutTests/fullscreen/fullscreen-containing-block-change-expected.txt: Added.
* LayoutTests/fullscreen/fullscreen-containing-block-change.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/media-controls.css:
(.media-controls-container):
* Source/WebCore/Modules/modern-media-controls/controls/vision-media-controls.css:
(.media-controls-container): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::addToTopLayer):
(WebCore::Element::removeFromTopLayer):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5198e14b51f73c112c7f28c80558a9e491190810

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12348 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12684 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14091 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11878 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12701 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14557 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12512 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14513 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10559 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11188 "7 failures") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18293 "2 flakes 119 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11642 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11353 "4 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14535 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11844 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9775 "1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11070 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11059 "Exiting early after 10 failures. 10 tests run. 1 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15400 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -509,10 +509,21 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
+        auto* newContainingBlock = renderer ? renderer->containingBlock() : nullptr;
+        if (oldContainingBlock != newContainingBlock) {
+            renderer->setNeedsLayout();
+            renderer->layout();
```
setNeedsLayout marks the containing block chain dirty and schedules a layout. is that the expected behavior here (judging by the next line (renderer->layout(), probably not)?

alanbaradlay commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -509,10 +509,21 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
+
+        // resolveStyle() will trigger layout if there are container queries, we need to invalidate
+        // the renderer again since the containing block could have since changed.
+        auto* newContainingBlock = renderer ? renderer->containingBlock() : nullptr;
```
is "renderer" guaranteed to be valid after remove/add (not sure what kind of activities they implicitly trigger like style resolve)?

anttijk commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -509,10 +509,22 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
+        if (oldContainingBlock != newContainingBlock) {
+            renderer->setChildNeedsLayout(MarkOnlyThis);
+            renderer->layout();
+        }
```
If the problem is the lack of correct layout invalidation then it would be better to do in removeFromTopLayer/addToTopLayer so other call sites would get this right too. Having to do setChildNeedsLayout manually indicates that the renderer tree dirty bits are in inconsistent state, can we avoid getting into such state in the first place?

Why is synchronous layout needed here? Calling layout on a single renderer and skipping the normal layout entry points seems sketchy.

nt1m committed:
REGRESSION(265043@main): Fullscreen mode only shows video in some part of the screen on Twitter
https://bugs.webkit.org/show_bug.cgi?id=258907
rdar://111095697

Reviewed by NOBODY (OOPS!).

265043@main started to use container queries in the video shadow root. As a side effect, this caused every call to
Document::resolveStyle() to start laying out the whole document (through frameView.layoutContext().layout()).

Adding to top layer & removing from top layer requires layout when the containing block changes.

We haven't really hit any problem with not marking for layout until now, because usually we apply the pseudo element styles
which include `position: fixed;` and other things, at the same time as adding to top layer, which marks them for layout anyway.

In the fullscreen case, we have a slightly different series:
```
ancestor->setFullscreenFlag(true); <- applies pseudo class styles (hence dirtying layout)

ancestor->document().resolveStyle(Document::ResolveStyleType::Rebuild); <- runs layout (clearing dirty layout)

ancestor->addToTopLayer(); <- changes the containing block without marking renderer for layout

// No layout!!
```

Add a special case in addToTopLayer/removeFromTopLayer, for cases where we don't process pseudo class style changes and containing block changes
under the same layout.

Also make it clear in the media-controls CSS that container queries are needed on all platforms (not just visionOS), since
WebVTT relies on it.

* LayoutTests/fullscreen/fullscreen-containing-block-change-expected.txt: Added.
* LayoutTests/fullscreen/fullscreen-containing-block-change.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/media-controls.css:
(.media-controls-container):
* Source/WebCore/Modules/modern-media-controls/controls/vision-media-controls.css:
(.media-controls-container): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::addToTopLayer):
(WebCore::Element::removeFromTopLayer):

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3868,6 +3871,16 @@ void Element::addToTopLayer()
+    // marks them for layout anyway.
+    renderer = this->renderer();
+    auto* newContainingBlock = renderer ? renderer->containingBlock() : nullptr;
+    if (oldContainingBlock != newContainingBlock && !renderer->needsLayout())
+        renderer->setNeedsLayout();
```
There is no to test for needsLayout(), setNeedsLayout() bails out if there is nothing to do (also in the other place).

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3868,6 +3871,16 @@ void Element::addToTopLayer()
+    // This requires a layout. In most cases, this isn't useful since we usually
+    // change pseudo-class styles at the same time as adding to the top layer, which
+    // marks them for layout anyway.
+    renderer = this->renderer();
+    auto* newContainingBlock = renderer ? renderer->containingBlock() : nullptr;
```
Caching the renderer to a local doesn't really do much. At least it could be a WeakPtr for safety.
